### PR TITLE
Feat/async api docs

### PR DESCRIPTION
**Pull request by lmoesle**
---
name: Feat/async api docs
about: Contribute your changes to the project
title: Feat/async api docs
labels: ''
assignees: ''
---
**Description**

Spring boot starter for auto-generation of async-api docs (kafka topics).

**Reference**
Issues [#144](https://github.com/it-at-m/digiwf-project/issues/144)

**Checklist**

- [ ] Acceptance criteria are met
- [x] Pipeline has been run successfully
- [x] JUnit tests have been run successfully
- [x] Changelog is adjusted
- [x] [Documentations](https://git.muenchen.de/groups/digitalisierung/-/wikis/Dokumentationen) are completed
- [ ] Frontend is tested
- [ ] Created sub-branches are deleted
- [ ] Boards are updated ( [Digitalisierung](https://git.muenchen.de/groups/digitalisierung/-/boards) , [Support](https://git.muenchen.de/digitalisierung/digiwf-support) , [Zenhub](https://app.zenhub.com/workspaces/digiwf-621f70bf50ea1100120b7e93/board) )
- [ ] feature process is created or extended
- [ ] [git-ops](https://git.muenchen.de/digitalisierung/digiwf-ops) is customized
- [ ] Openshift environments are prepared (Secrets, etc.)

### CODE REVIEW

**Comment by ghost** [comment]
Die [Dokumentation Repositories](https://git.muenchen.de/groups/digitalisierung/-/wikis/Dokumentationen) fehlt noch.

File: digiwf-asyncapi-docs-core/src/main/java/io/muenchendigital/digiwf/AsyncApiConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        final Server kafkaServer = Server.builder()
+                .protocol(this.binder)
+                .url(this.broker)
+                .build();
+
+        // registering producers is not required. Producers and consumers are automatically detected.
+        return AsyncApiDocket.builder()
+                .basePackage(this.basePackage)
+                .info(info)
+                .server("kafka", kafkaServer)
```
Comment (dominikhorn93) [inline digiwf-asyncapi-docs-core/src/main/java/io/muenchendigital/digiwf/AsyncApiConfiguration.java]: Is there a reason why kafka is hard coded here?

File: digiwf-asyncapi-docs-core/src/main/java/io/muenchendigital/digiwf/DocumentAsyncAPI.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+import java.lang.annotation.Target;
+
+/**
+ * Automatically document spring cloud stream producer and consumer functions that reside in a spring component.
+ * Therefore, annotate the class with @DocumentAsyncAPI(payload = Payload.class).
+ *
+ * Restrictions: This annotation only works for regular producer functions. If you use dynamic routing,
+ * you have to specify producers manually.
+ */
+@Target({ElementType.TYPE})
```
Comment (dominikhorn93) [inline digiwf-asyncapi-docs-core/src/main/java/io/muenchendigital/digiwf/DocumentAsyncAPI.java]: should it be possible on method level? We can have multiple consumers in one class

File: digiwf-asyncapi-docs-core/src/main/java/io/muenchendigital/digiwf/AsyncApiConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    private String version;
+    private String title;
+
+    public AsyncApiDocket asyncApiDocket() {
+        final Info info = Info.builder()
+                .version(this.version)
+                .title(this.title)
+                .build();
+
+        final Server kafkaServer = Server.builder()
```
Comment (dominikhorn93) [inline digiwf-asyncapi-docs-core/src/main/java/io/muenchendigital/digiwf/AsyncApiConfiguration.java]: final Server server?

File: digiwf-asyncapi-docs-starter/src/main/java/io/muenchendigital/digiwf/configuration/AsyncAPIDocsAutoConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        return new ConsumerAndProducerScanner(
+                schemasService,
+                this.definitionProperties.getDefinitions(),
+                this.bindingProperties.getBindings(),
+                this.docsProperties.getBasePackage()
+        );
+    }
+
+    @Bean
+    @ConditionalOnProperty(name = "spring.cloud.stream.function.routing.enabled")
```
Comment (dominikhorn93) [inline digiwf-asyncapi-docs-starter/src/main/java/io/muenchendigital/digiwf/configuration/AsyncAPIDocsAutoConfiguration.java:48]: Good job!

**Comment by dominikhorn93** [comment]
In my point of view it only works with kafka right? There are a lot of kakfa related dependencies and classes and dependencies. Should this be more present in the README?

**Review by dominikhorn93 — APPROVED** [review]

**Pull request merged**